share/spack/spack-completion.bash

```diff
@@ -80,1266 +104,1448 @@ _bash_completion_spack () {
     else
-        compgen -W "$(_all_packages)" -- "$cur"
+        _all_packages
     fi
 }
 
-_spack_env () {
+_spack_env() {
     if $list_options
     then
-        compgen -W "-h --help" -- "$cur"
+        SPACK_COMPREPLY="-h --help"
     else
-        compgen -W "activate deactivate create remove rm list ls status st loads view" -- "$cur"
+        SPACK_COMPREPLY="activate deactivate create remove rm list ls status st loads view"
     fi
 }
 
-_spack_env_activate () {
+_spack_env_activate() {
     if $list_options
     then
-        compgen -W "-h --help --sh --csh -v --with-view -V --without-view -d --dir -p --prompt" -- "$cur"
+        SPACK_COMPREPLY="-h --help --sh --csh -v --with-view -V --without-view -d --dir -p --prompt"
     else
-        compgen -W "$(_environments)" -- "$cur"
+        _environments
     fi
 }
 
-_spack_env_deactivate () {
-    compgen -W "-h --help --sh --csh" -- "$cur"
+_spack_env_deactivate() {
+    SPACK_COMPREPLY="-h --help --sh --csh"
 }
 
-_spack_env_create () {
+_spack_env_create() {
     if $list_options
     then
-        compgen -W "-h --help -d --dir --without-view --with-view" -- "$cur"
+        SPACK_COMPREPLY="-h --help -d --dir --without-view --with-view"
+    else
+        _environments
     fi
 }
 
-_spack_env_remove () {
+_spack_env_remove() {
     if $list_options
     then
-        compgen -W "-h --help -y --yes-to-all" -- "$cur"
+        SPACK_COMPREPLY="-h --help -y --yes-to-all"
     else
-        compgen -W "$(_environments)" -- "$cur"
+        _environments
     fi
 }
 
-_spack_env_rm () {
-    # Alias to `spack env remove`
-    _spack_env_remove
+_spack_env_rm() {
+    if $list_options
+    then
+        SPACK_COMPREPLY="-h --help -y --yes-to-all"
+    else
+        _environments
+    fi
 }
 
-_spack_env_list () {
-    compgen -W "-h --help" -- "$cur"
+_spack_env_list() {
+    SPACK_COMPREPLY="-h --help"
 }
 
-_spack_env_ls () {
-    # Alias to `spack env list`
-    _spack_env_list
+_spack_env_ls() {
+    SPACK_COMPREPLY="-h --help"
 }
 
-_spack_env_status () {
-    compgen -W "-h --help" -- "$cur"
+_spack_env_status() {
+    SPACK_COMPREPLY="-h --help"
 }
 
-_spack_env_st () {
-    # Alias to `spack env status`
-    _spack_env_status
+_spack_env_st() {
+    SPACK_COMPREPLY="-h --help"
 }
 
-_spack_env_loads () {
+_spack_env_loads() {
     if $list_options
     then
-        compgen -W "-h --help -m --module-type --input-only -p --prefix -x --exclude -r --dependencies" -- "$cur"
+        SPACK_COMPREPLY="-h --help -m --module-type --input-only -p --prefix -x --exclude -r --dependencies"
     else
-        compgen -W "$(_environments)" -- "$cur"
+        _environments
     fi
 }
 
-_spack_env_view () {
+_spack_env_view() {
     if $list_options
     then
-        compgen -W "-h --help" -- "$cur"
+        SPACK_COMPREPLY="-h --help"
     else
-        compgen -W "regenerate enable disable" -- "$cur"
+        SPACK_COMPREPLY=""
```

Comment:
This is something I still need to fix. The actual positional arg is `{regenerate,enable,disable}`. If we reimplement this using subcommands, the problem goes away.